feat(ssr): tolerate circular imports

aleclarson opened the pull request:
<!-- Thank you for contributing! -->

### Description

This PR makes SSR imports work like Node.js in the following way:

Modules that depend on each other will never receive an undefined exports. Instead, the module that is first to import the other module will wait for the other module to be instantiated. When the other module imports the first module, it receives the incomplete exports object, which will be completed when the first module is done being instantiated.

To accomplish this behavior, we need to load SSR dependencies **one at a time**. This ensures a predictable import order. Before this PR, the import order is non-deterministic, as it depends on the depth of the dependency tree plus the transformation time per module.

This PR also avoids infinite hangs caused by a dependency having its `ssrLoadModule` promise awaited without first checking if one of its dependencies is within the `urlStack`.

As of 0333312, this PR also allows the "other module" to access the partial exports of the "first module". For example, if `export { foo } from "./foo"` comes before `export { bar } from "./bar"` in the `./index.js` module, then `./bar.js` will still be able to `import { foo } from "./index"` and use it before `./index.js` has finished being instantiated.

### Todo

- [ ] Split out the `playground/ssr-react` changes into a new playground `ssr-circular-import`

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

Fixes #2258
Fixes #2491 

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [ ] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

fairbanksg committed:
fix: deadlock in ssrLoadModule with circular dependencies (fix #2491, fix #2258)

aleclarson committed:
feat(ssr): tolerate circular imports

aleclarson committed:
fix(ssr): load dependencies in predictable order

If dependencies are not loaded one at a time, the load order depends on the depth of the dependency tree plus the transformation time per module. When forced to load serially, the load orders depends on the order of import declarations, which is how NodeJS works too.

aleclarson committed:
chore: revert ssrModuleLoader changes from #3473

raythurnvoid committed:
fix: improve circular dependency module initialization

raythurnevoid commented:
@aleclarson i've brought my improve to your branch: https://github.com/aleclarson/vite/pull/2

in can still be improved but it works on the project i'm working at the moment

aleclarson committed:
chore: revert ssrModuleLoader changes from raythurnevoid

aleclarson committed:
feat(ssr): await imports and co-locate exports

SSR modules are now wrapped in an async function, so they can await their imports. Also, exports are now always defined in place, instead of being appended to the bottom of the module. This allows better parity with how ESM handles circular imports.

aleclarson commented:
@raythurnevoid Thanks for the test case. I've solved it in a better way, see here: https://github.com/vitejs/vite/pull/3950/commits/0333312535f3069d53d2a0e746931fcf0cde020e

raythurnevoid commented:
> @raythurnevoid Thanks for the test case. I've solved it in a better way, see here: [0333312](https://github.com/vitejs/vite/commit/0333312535f3069d53d2a0e746931fcf0cde020e)

I can confirm it is working on the project i'm working too

aleclarson committed:
refactor: skip unnecessary `isExternal` call

aleclarson committed:
refactor: inline `isExternal` call

aleclarson committed:
fix(ssr): account for multiple pending dynamic imports

aleclarson committed:
fix(ssr): look for dep in `urlStack`

patak-js commented:
@aleclarson would you resolve the conflicts? Is this one ready to be discussed with the team?

aleclarson committed:
chore: fix lint

Shinigami92 commented:
Should @brillout have a look?

brillout commented:
@Shinigami92 LGTM 

@aleclarson Neat neat - one thing though: how about we create a new playground instead of populating `playground/ssr-react/`? I often point people to `playground/ssr-react` and `playground/ssr-vue` to show them Vite's native SSR API. Would be good for them if we keep these playgrounds minimal. Thoughts?

@LydiaF Hope this makes you re-consider Vite hehe.

LydiaF commented:
> @Shinigami92 LGTM (Although my review is superficial and I didn't dig that much)
> 
> @aleclarson Neat neat - one thing though: how about we create a new playground instead of populating `playground/ssr-react/`? I often point people to `playground/ssr-react` and `playground/ssr-vue` to show them Vite's native SSR API. Would be good for them if we keep these playgrounds minimal. Thoughts?
> 
> @LydiaF Hope this makes you re-consider Vite hehe.

@brillout we never left vite! We just left ssr for a bit and just coincidentally, today we are giving ssr another go!

brillout commented:
> @brillout we never left vite! We just left ssr for a bit and just coincidentally, today we are giving ssr another go!

Ah nice, keep me updated about how SSR works out for you. Many people have been using vite-plugin-ssr + GraphQL Apollo in the meantime. Also check out https://vite-plugin-ssr.com/path-aliases it works now.

Make sure to note that:

> Vite's vite.config.js#resolve.alias only works for Vite processed files (all .page.js / .page.server.js / .page.client.js / .page.route.js files and all their imports). For files that are not processed by Vite (e.g. Express.js server code) we can use a Node.js aliasing tools such as module-alias (all browser files are always processed by Vite).

aleclarson commented:
> how about we create a new playground instead of populating `playground/ssr-react/`? I often point people to `playground/ssr-react` and `playground/ssr-vue` to show them Vite's native SSR API. Would be good for them if we keep these playgrounds minimal. Thoughts?

I agree. The tests were contributed by others. If someone can create the new playground, it would be appreciated. 👍

Shinigami92 commented on packages/vite/src/node/ssr/ssrModuleLoader.ts:
```diff
@@ -46,7 +40,13 @@ export async function ssrLoadModule(
   pendingModules.set(url, modulePromise)
-  modulePromise.catch(() => {}).then(() => pendingModules.delete(url))
+  modulePromise
+    .catch(() => {
+      pendingImports.delete(url)
+    })
+    .then(() => {
+      pendingModules.delete(url)
+    })
```
**suggestion:** What about merging this into one `finally`?

aleclarson committed:
chore: use promise.finally

patak-js commented on packages/vite/src/node/ssr/ssrModuleLoader.ts:
```diff
@@ -46,7 +40,13 @@ export async function ssrLoadModule(
   pendingModules.set(url, modulePromise)
-  modulePromise.catch(() => {}).then(() => pendingModules.delete(url))
+  modulePromise
+    .catch(() => {
+      pendingImports.delete(url)
+    })
```
This is no longer needed now, no?
```suggestion
```

aleclarson replied:
Still needed, since `instantiateModule` promise will reject when a `ssrImport` promise is rejected, which happens after `pendingImports` is set but before it's cleaned up.

patak-js replied:
Oh... sorry for the noise then. I was confused by this comment https://github.com/vitejs/vite/pull/3950#discussion_r680722085, I thought that both bodies of catch and then were equal. Then there isn't a difference in changing the then to finally.

aleclarson replied:
Right, just more semantically clear.

rtsao commented:
It seems like there is still possible issues with circular imports and SSR that can lead to intermittent hanging: https://github.com/vitejs/vite/issues/11468
